packages/vuetify/src/components/VDataTable/composables/expand.ts

```diff
@@ -40,24 +40,23 @@ export function provideExpanded (props: ExpandProps) {
     return [...v.values()]
   })
 
-  function getItemKey (item: DataTableItem) {
-    return isObject(item.value) ? item.key : item.value
-  }
-
   function expand (item: DataTableItem, value: boolean) {
     const newExpanded = new Set(expanded.value)
+    const rawValue = toRaw(item.value)
 
     if (!value) {
-      newExpanded.delete(getItemKey(item))
+      const item = [...expanded.value].find(x => toRaw(x) === rawValue)!
```

Comment:
Not a huge fan of this but expanded shouldn't contain too many items so it's probably ok.